Adds rejoin system

Itsnotjames opened the pull request:
Removes HandleDisconnect() and it's call in CustomNetworkManager.cs.
Moves part of RespawnPlayer() to a new function: TransferPlayer().
Adds List<ConnectedPlayer> loggedOff in PlayerList.cs. When players are removed in TryRemove() they are added to the LoggedOff list to be retrieved when they relog.
Added TakeLoggedOffPlayer() in PlayerList.cs which returns a logged player with a certain steamId and removes them from the list.
Refactored JoinedViewer's player setup so the client sends a steamId to the server, the server does serverside player setup with the steamId, then tells the client to do clientside player setup with the character they logged out with (if they logged out).
Added CmdRejoin() in JoinedViewer which lets the player rejoin their character.
Adds ReenterBodyUpdates(), SendSyncMessage(), and UpdateInventorySlots() to PlayerNetworkActions.cs, which are used to sync a rejoined player.
Creates setLocalPlayer() in PlayerSync.cs from existing setLocalPlayer code so it can be used with the rejoin system.
Stops the player gameobject from being destroyed on logout.
Gives CmdServerSetupPlayer() "0" (ConnectedPlayer.steamId's default value) as an argument if it is not a steam server.

### Purpose
Lets players rejoin their character after logging out and logging back in.
Fixes #1429 

### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [X]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [X]  This PR has been tested in multiplayer (with 2 clients, if applicable)

### Notes:
Should be tested on steam to make sure Client.Instance.SteamId works.

Itsnotjames committed:
Adds rejoin system

Removes HandleDisconnect() and it's call in CustomNetworkManager.cs.
Moves part of RespawnPlayer() to a new function: TransferPlayer().
Adds List<ConnectedPlayer> loggedOff in PlayerList.cs. When players are removed in TryRemove() they are added to the LoggedOff list to be retrieved when they relog.
Added TakeLoggedOffPlayer() in PlayerList.cs which returns a logged player with a certain steamId and removes them from the list.
Refactored JoinedViewer's player setup so the client sends a steamId to the server, the server does serverside player setup with the steamId, then tells the client to do clientside player setup with the character they logged out with (if they logged out).
Added CmdRejoin() in JoinedViewer which lets the player rejoin their character.
Adds ReenterBodyUpdates(), SendSyncMessage(), and UpdateInventorySlots() to PlayerNetworkActions.cs, which are used to sync a rejoined player.
Creates setLocalPlayer() in PlayerSync.cs from existing setLocalPlayer code so it can be used with the rejoin system.
Stops the player gameobject from being destroyed on logout.
Gives CmdServerSetupPlayer() "0" (ConnectedPlayer.steamId's default value) as an argument if it is not a steam server.

chairbender commented on UnityProject/Assets/Scripts/Managers/NetworkManagement/SpawnHandler.cs:
```diff
@@ -114,4 +108,16 @@ private static Transform GetSpawnForJob(JobType jobType)
 		return spawnPoints.Count == 0 ? null : spawnPoints.PickRandom().transform;
 	}
+
+	public static void TransferPlayer(NetworkConnection conn, short playerControllerId, GameObject player){
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/JoinedViewer.cs:
```diff
@@ -85,4 +122,11 @@ public void CmdRequestJob(JobType jobType)
     }
+
+	[Command]
+	public void CmdRejoin(GameObject loggedOffPlayer)
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -86,7 +88,16 @@ public override void OnStartServer()
 	}
 
+	[Server]
+	public void ReenterBodyUpdates(GameObject recipient){
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -86,7 +88,16 @@ public override void OnStartServer()
+	}
 
+	[Server]
+	public void SendSyncMessage(GameObject recipient){
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/PlayerNetworkActions.cs:
```diff
@@ -314,6 +325,16 @@ public void SetInventorySlot(string slotName, GameObject obj)
 	}
 
+	[Server]
+	public void UpdateInventorySlots(GameObject recipient)
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/PlayerSync.cs:
```diff
@@ -376,6 +374,12 @@ private void Start()
 		pushPull = GetComponent<PushPull>();
 	}
 
+	public void setLocalPlayer(){
```
Please add /// documentation to this

chairbender commented on UnityProject/Assets/Scripts/Player/PlayerSync.Client.cs:
```diff
@@ -21,7 +21,7 @@ public Vector3IntEvent OnClientTileReached()
 	private PlayerState predictedState;
 
-	private Queue<PlayerAction> pendingActions;
+	public Queue<PlayerAction> pendingActions;
```
Please keep this private. It should work fine to keep it private even though the code is in a different .cs file because it's still in the same partial class.

chairbender reviewed (changes requested):
Great stuff!

Just a few minor changes - want to get some documentation on those new public methods among other things.

Other than that...

I notice that the original ticket is asking for the Brain GUID to be used to figure out which player to rejoin as, but here you are using SteamID, which is tied to a specific ConnectedPlayer. The intention for tying it to Brain GUID is so that a player could have their brain transplanted into something - such as another player, a robot, a monkey, etc...But I think this can still be achieved using ConnectedPlayer (just shift the connected player to whatever their brain is moved to), and most people in Discord / comments on the ticket seem to think it's fine to use SteamID for now. So I'm going to say it's fine to merge as is - we can handle brain swapping when we get to it.

Another possible issue is that it's based on SteamID, meaning it wouldn't work if attempting to run a non-steam version of the game, as you already discussed in Discord. When we want to support a different way of doing it, I think it should be as simple as changing ConnectedPlayer to add some other ID to it. So I think it's fine to merge as-is.

Itsnotjames committed:
Adds summaries and small fix

Adds summaries to: SpawnHandler.TransferPlayer(), JoinedViewer.CmdRejoin(), PlayerNetworkActions.ReenterBodyUpdates(), PlayerNetworkActions.SendSyncMessage(), PlayerNetworkActions.UpdateInventorySlots(), and PlayerSync.setLocalPlayer().
Makes PlayerSync.Client.pendingActions private again.
